From cad985799a0fd644002fd75795e156894f256b43 Mon Sep 17 00:00:00 2001
From: Chih-Wei Huang <cwhuang@linux.org.tw>
Date: Mon, 20 Apr 2020 18:43:48 +0800
Subject: [PATCH 1/2] Make sleep state configurable and add a fallback (s-x86)

This patch allows the user to set the sleep state target from
Android properties for wakeup_count method.

It also includes a fallback state if the default state is not
available and the user didn't set the sleep.state property.

Changelog:
(r-x86) Resolve conflict in suspend/1.0/default/SystemSuspend.cpp
(s-x86) Resolve conflict in suspend/1.0/default/SystemSuspend.cpp

Change-Id: I74e61b3f77cbded24578a8f9463dc4c6f3b660a7
---
 suspend/1.0/default/SystemSuspend.cpp | 25 +++++++++++++++++++++++--
 suspend/1.0/default/SystemSuspend.h   |  2 ++
 2 files changed, 25 insertions(+), 2 deletions(-)

diff --git a/suspend/1.0/default/SystemSuspend.cpp b/suspend/1.0/default/SystemSuspend.cpp
index e9bb646..8cd0d92 100644
--- a/suspend/1.0/default/SystemSuspend.cpp
+++ b/suspend/1.0/default/SystemSuspend.cpp
@@ -18,6 +18,7 @@
 
 #include <android-base/file.h>
 #include <android-base/logging.h>
+#include <android-base/properties.h>
 #include <android-base/stringprintf.h>
 #include <android-base/strings.h>
 #include <fcntl.h>
@@ -30,6 +31,7 @@
 #include <thread>
 
 using ::android::base::Error;
+using ::android::base::GetProperty;
 using ::android::base::ReadFdToString;
 using ::android::base::WriteStringToFd;
 using ::android::hardware::Void;
@@ -194,7 +196,7 @@ bool SystemSuspend::forceSuspend() {
     //  returns from suspend, the wakelocks and SuspendCounter will not have
     //  changed.
     auto counterLock = std::unique_lock(mCounterLock);
-    bool success = WriteStringToFd(kSleepState, mStateFd);
+    bool success = WriteStringToFd(getSleepState(), mStateFd);
     counterLock.unlock();
 
     if (!success) {
@@ -269,7 +271,7 @@ void SystemSuspend::initAutosuspend() {
                 PLOG(VERBOSE) << "error writing from /sys/power/wakeup_count";
                 continue;
             }
-            bool success = WriteStringToFd(kSleepState, mStateFd);
+            bool success = WriteStringToFd(getSleepState(), mStateFd);
             counterLock.unlock();
 
             if (!success) {
@@ -295,6 +297,25 @@ void SystemSuspend::initAutosuspend() {
     LOG(INFO) << "automatic system suspend enabled";
 }
 
+const string &SystemSuspend::getSleepState() {
+    if (mSleepState.empty()) {
+        mSleepState = GetProperty("sleep.state", "");
+        if (!mSleepState.empty()) {
+            LOG(INFO) << "autosuspend using sleep.state property " << mSleepState;
+        } else {
+            string buf = readFd(mStateFd);
+            if (buf.find(kSleepState) != std::string::npos) {
+                mSleepState = kSleepState;
+                LOG(INFO) << "autosuspend using default sleep_state " << mSleepState;
+            } else {
+                mSleepState = "freeze";
+                LOG(WARNING) << "autosuspend using fallback state " << mSleepState;
+            }
+        }
+    }
+    return mSleepState;
+}
+
 /**
  * Updates sleep time depending on the result of suspend attempt.
  * Time (in milliseconds) between suspend attempts is described the formula
diff --git a/suspend/1.0/default/SystemSuspend.h b/suspend/1.0/default/SystemSuspend.h
index b84b1c9..4cbc218 100644
--- a/suspend/1.0/default/SystemSuspend.h
+++ b/suspend/1.0/default/SystemSuspend.h
@@ -118,12 +118,14 @@ class SystemSuspend : public ISystemSuspend {
 
    private:
     void initAutosuspend();
+    const std::string &getSleepState();
 
     std::mutex mCounterLock;
     std::condition_variable mCounterCondVar;
     uint32_t mSuspendCounter;
     unique_fd mWakeupCountFd;
     unique_fd mStateFd;
+    std::string mSleepState;
 
     unique_fd mSuspendStatsFd;
     unique_fd mSuspendTimeFd;
-- 
2.25.1

